RecoMET/METAlgorithms/src/CSCHaloAlgo.cc

```diff
@@ -757,44 +639,695 @@ math::XYZPoint CSCHaloAlgo::getPosition(const DetId &id, reco::Vertex::Point vtx
-    if(
-       (rechit.time()<-3)&&
-       (rhpos.z()*iZ<0|| abs(rhpos.z())<200)&&
-       rhet> et_thresh_rh&&
-       dphi_rhseg < dphi_thresh_segvsrh &&
-       dr_rhseg < dr_highthresh_segvsrh && dr_rhseg> dr_lowthresh_segvsrh && //careful: asymmetric cut might not be the most appropriate thing 
-       dtcorr_rhseg> dt_lowthresh_segvsrh && dtcorr_rhseg< dt_highthresh_segvsrh
-      ) return true; 
+    if(rhet<et_thresh_seedrh) continue;
+    double eta = rhpos.eta();
+    double phi = rhpos.phi();
+    
+    bool isiso = true;
+    double etcluster(0);
+    int nbcrystalsameeta(0);
+    double timediscriminator(0);
+    double etstrip_iphiseedplus1(0), etstrip_iphiseedminus1(0);
+
+    //Building the cluster 
+    edm::RefVector<EcalRecHitCollection> bhrhcandidates;
+    for(size_t jhit = 0; jhit<ecalrechitcoll->size(); ++ jhit){
+      const EcalRecHit & rechitj = (*ecalrechitcoll)[ jhit ];
+      EcalRecHitRef rhRef(ecalrechitcoll,jhit);
+      math::XYZPoint rhposj = getPosition(rechitj.id(),vtx);
+
+      double etaj = rhposj.eta();
+      double phij = rhposj.phi();
+      if(abs(eta-etaj)>0.2) continue;//This means +/-11 crystals in eta 
+      if(abs(deltaPhi(phi,phij))>0.08) continue;//This means +/-4 crystals in phi 
+      
+      double rhetj = rechitj.energy()/cosh(rhposj.eta());
+      //Rechits with et between 1 and 2 GeV are saved in the rh list but not used in the calculation of the halocluster variables
+      if(rhetj<1) continue;
+      bhrhcandidates.push_back(rhRef);
+      if(rhetj<2) continue;
+
+      if(abs(deltaPhi(phi,phij))>0.03){isiso=false;break;}//The strip should be isolated
+      if(abs(deltaPhi(phi,phij))<0.01) nbcrystalsameeta++;
+      if(deltaPhi(phi,phij)>0.01) etstrip_iphiseedplus1+=rhetj;
+      if(deltaPhi(phi,phij)<-0.01) etstrip_iphiseedminus1+=rhetj;
+      etcluster+=rhetj;
+      //Timing discriminator
+      //We assign a weight to the rechit defined as: 
+      //Log10(Et)*f(T,R,Z) 
+      //where f(T,R,Z) is the separation curve between halo-like and IP-like times.
+      //The time difference between a deposit from a outgoing IT halo and a deposit coming from a particle emitted at the IP is given by: 
+      //dt= ( - sqrt(R^2+z^2) + |z| )/c
+      //Here we take R to be 130 cm. 
+      //For EB, the function was parametrized as a function of ieta instead of Z.
+      double rhtj = rechitj.time();
+      EBDetId detj    = rechitj.id();
+      int rhietaj= detj.ieta();
+      timediscriminator+= TMath::Log10( rhetj )* ( rhtj +0.5*(sqrt(16900+9*rhietaj*rhietaj)-3*abs(rhietaj))/30 );
+      
+    }
+    //Isolation condition
+    if(!isiso) continue;
+    
+    //Calculate H/E
+    double hoe(0);
+    for(size_t jhit = 0; jhit<hbherechitcoll->size(); ++ jhit){
+      const HBHERecHit & rechitj = (*hbherechitcoll)[ jhit ];
+      math::XYZPoint rhposj = getPosition(rechitj.id(),vtx);
+      double rhetj = rechitj.energy()/cosh(rhposj.eta());
```

Comment:
``` sqrt(rhposj.mag2()/rhposj.perp2()) ``` is a less numerically challenging operation than a chain of exp,log,tan,atan used in cosh(eta())